### PR TITLE
[WIP] run: add --outs-persist and --outs-persist-no-cache options

### PR DESCRIPTION
**Pull request by pared**
Fixes to #1214

### CODE REVIEW

File: tests/test_data_cloud.py [hunk 1]
```diff
@@ -628,8 +628,7 @@ def _test(self):
         url = get_local_url()
         self.main(["remote", "add", "-d", TEST_REMOTE, url])
 
-        self.dvc.add(self.FOO)
-        stage = self.dvc.run(deps=["foo"], outs=["bar"], cmd="echo bar > bar")
+        stage = self.dvc.run(outs=["bar"], cmd="echo bar > bar")
```
Comment (efiop) [inline tests/test_data_cloud.py]: How is this related to the patch?

File: tests/test_run.py [hunk 2]
```diff
@@ -768,3 +768,25 @@ def test(self):
...
+        file = "file.txt"
+        file_content = "content"
+        stage_file = file + Stage.STAGE_FILE_SUFFIX
+
+        ret = main(
+            [
+                "run",
+                "-o",
+                file,
+                "--outs-no-remove",
```
Comment (efiop) [inline tests/test_run.py]: `--outs-no-remove` should work the same as existing `--outs` and `--outs-no-cache`.

Comment (efiop) [reply tests/test_run.py]: So `dvc run --outs-no-remove foo` should be used and not `dvc run --outs foo --outs-no-remove foo`. I like the idea of actually marking them post-factum, but for now we should stay in sync with existing behaviour.

File: tests/test_data_cloud.py (hunk 1 above)
Comment (pared) [reply tests/test_data_cloud.py]: Well, its not that much, however this test was failing due to logic in Repo._collect_used_cache. I decided to simplify this test. DO you want me to revert it?

Comment (efiop) [reply tests/test_data_cloud.py]: @pared Yes, let's at least not touch that one silently within this PR :slightly_smiling_face:

File: dvc/stage.py [hunk 3]
```diff
@@ -257,7 +258,13 @@ def remove_outs(self, ignore_remove=False):
         Used mainly for `dvc remove --outs`
         """
         for out in self.outs:
-            out.remove(ignore_remove=ignore_remove)
+            if (
+                OutputLOCAL.PARAM_REMOVE in out.info.keys()
```
Comment (efiop) [inline dvc/stage.py]: Please use `out.remove` or `out.persist` property.

File: dvc/repo/__init__.py [hunk 4]
```diff
@@ -195,11 +195,14 @@ def _collect_dir_cache(
...
 
+    def _out_missing_version_info(self, out):
+        return self.cache.local.PARAM_CHECKSUM not in out.info.keys()
+
     def _collect_used_cache(
         self, out, branch=None, remote=None, force=False, jobs=None
     ):
-        if not out.use_cache or not out.info:
-            if not out.info:
+        if not out.use_cache or self._out_missing_version_info(out):
```
Comment (efiop) [inline dvc/repo/__init__.py]: You should pop PARAM_REMOVE here https://github.com/iterative/dvc/blob/master/dvc/output/__init__.py#L75 and then you won't need to modify this.

File: dvc/stage.py [hunk 5]
```diff
@@ -260,7 +260,10 @@ def remove_outs(self, ignore_remove=False):
         Used mainly for `dvc remove --outs`
         """
         for out in self.outs:
-            out.remove(ignore_remove=ignore_remove)
+            if out.persist:
+                self.repo.unprotect(out.path)
```
Comment (efiop) [inline dvc/stage.py]: This will lead to a bug in non-local out case. What we need to do is implement remote.unprotect and out.unprotect(that would call remote.unprotect). And also use remote.unprotect in repo.unprotect.

File: dvc/output/base.py [hunk 6]
```diff
@@ -231,3 +241,6 @@ def get_files_number(self):
             return 1
 
         return 0
+
+    def unprotect(self):
+        self.remote.unportect(self.path)
```
Comment (efiop) [inline dvc/output/base.py]: Typo.

File: dvc/remote/local.py [hunk 7]
```diff
@@ -743,3 +744,6 @@ def _log_missing_caches(self, checksum_info_dict):
                 "nor on remote. Missing cache files: {}".format(missing_desc)
             )
             logger.warning(msg)
+
+    def unportect(self, target):
+        unprotect(target)
```
Comment (efiop) [inline dvc/remote/local.py]: This is backwards. Move repo.unprotect logic to RemoteLOCAL.unprotect and use it.

Comment (pared) [reply dvc/remote/local.py]: Done

Comment (efiop) [reply dvc/remote/local.py]: Please move logic from `repo.unprotect()` to `RemoteLOCAL.unprotect()` and use `RemoteLOCAL.unprotect` in `repo.unprotect()`.

File: dvc/output/base.py [hunk 8]
```diff
@@ -231,3 +241,7 @@ def get_files_number(self):
             return 1
 
         return 0
+
+    def unprotect(self):
+        if self.exists:
+            self.remote.unprotect(self.path)
```
Comment (efiop) [inline dvc/output/base.py]: Should be
```suggestion
            self.remote.unprotect(self.path_info)
```
with all the logical conequenses 🙂

File: dvc/stage.py [hunk 9]
```diff
@@ -262,18 +262,19 @@ def changed(self):
...
+                        out=out, stage=self.relpath
+                    )
                 )
-            )
-            out.remove(ignore_remove=ignore_remove)
+                out.remove(ignore_remove=ignore_remove)
 
     def unprotect_outs(self):
         for out in self.outs:
-            if out.scheme != "local" or not out.exists:
```
Comment (efiop) [inline dvc/stage.py]: Previously we were simply ignoring an output if it didn't exist, but now `unprotect()` will raise an exception. This  will break `dvc repro` when output doesn't exist.

Comment (pared) [reply dvc/stage.py]: This functionality still exists, I moved it to output base class.

Comment (efiop) [reply dvc/stage.py]: Ah, indeed! Sorry, didn't catch that :slightly_smiling_face:

**Comment by jorgeorpinel** [comment]
I just realize this was never documented 😋 iterative/dvc.org/issues/2027

**Comment by shcheklein** [comment]
There was a PR, we didn't merge it to keep this option hidden for now.

**Comment by jorgeorpinel** [comment]
Wait I was wrong. This is documented in https://dvc.org/doc/command-reference/run#options, it's just missing from dvc.yaml (explain well what "persist" means).
